x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedKeywordFieldMapper.java

```diff
@@ -0,0 +1,393 @@
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.io.stream.ByteArrayStreamInput;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.index.fielddata.AbstractSortedSetDocValues;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.FieldDataContext;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.LeafOrdinalsFieldData;
+import org.elasticsearch.index.fielddata.plain.AbstractIndexOrdinalsFieldData;
+import org.elasticsearch.index.fielddata.plain.AbstractLeafOrdinalsFieldData;
+import org.elasticsearch.index.mapper.BinaryFieldMapper;
+import org.elasticsearch.index.mapper.DocumentParserContext;
+import org.elasticsearch.index.mapper.FieldMapper;
+import org.elasticsearch.index.mapper.KeywordFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.Mapper;
+import org.elasticsearch.index.mapper.MapperBuilderContext;
+import org.elasticsearch.index.mapper.SourceValueFetcher;
+import org.elasticsearch.index.mapper.StringFieldType;
+import org.elasticsearch.index.mapper.TextSearchInfo;
+import org.elasticsearch.index.mapper.ValueFetcher;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.script.field.KeywordDocValuesField;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortOrder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static org.elasticsearch.common.lucene.Lucene.KEYWORD_ANALYZER;
+
+/**
+ * <p>A special field mapper for multi-valued keywords that may contain duplicate values. If the associated <code>counted_terms</code>
+ * aggregation is used, duplicates are considered in aggregation results. Consider the following values:</p>
+ *
+ * <ul>
+ *     <li><code>["a", "a", "b"]</code></li>
+ *     <li><code>["a", "b", "b"]</code></li>
+ * </ul>
+ *
+ * <p>While a regular <code>keyword</code> and the corresponding <code>terms</code> aggregation deduplicates values and reports a count of
+ * 2 for each key (one per document), a <code>counted_terms</code> aggregation on a <code>counted_keyword</code> field will consider
+ * the actual count and report a count of 3 for each key.</p>
+ *
+ * <p>Only regular source is supported; synthetic source won't work.</p>
+ */
+public class CountedKeywordFieldMapper extends FieldMapper {
```

Comment:
Maybe add a note here that this field mapper and associated aggregation are both an internal experimental feature and therefor not publicly documented and has the tech preview status?

Reply:
Sure. How about I'll leave the comment on the plugin which is the entry point to both?

Reply:
👍 That is good with me too.

Reply:
I've pushed the Javadoc comment in https://github.com/elastic/elasticsearch/pull/101826/commits/0162d9dd4e55e7966ab75818230ce1dfd7f0458d.

Reply:
👍